### PR TITLE
Support dynamic setting of memory

### PR DESCRIPTION
**Pull request by joii2020**
Support dynamic setting of memory.
completed:interpreter, asm_x64
todo: asm_aarch64

### CODE REVIEW

File: definitions/src/asm.rs [hunk 1]
```diff
@@ -48,20 +48,28 @@ pub struct AsmCoreMachine {
...
+        assert_ne!(memory_size, 0);
+        assert_eq!(memory_size % RISCV_PAGESIZE, 0);
+
         let mut machine = unsafe {
-            let layout = Layout::new::<AsmCoreMachine>();
-            #[allow(clippy::cast_ptr_alignment)]
-            // Use alloc so we are using malloc instead of
-            // calloc, then do lazy zero filling when necessary. That might
-            // save us some time in case a script doesn't use all the memory.
+            let machine_size =
```
Comment (XuJiandong) [inline definitions/src/asm.rs:73]: Add some comments here.

File: examples/ckb-vm-runner.rs [hunk 2]
```diff
@@ -116,9 +119,11 @@ fn main_int(code: Bytes, args: Vec<Bytes>) -> Result<(), Box<dyn std::error::Err
         ckb_vm::ISA_IMC | ckb_vm::ISA_B | ckb_vm::ISA_MOP,
         ckb_vm::machine::VERSION1,
         u64::MAX,
+        RISCV_MAX_MEMORY,
     );
     let machine_builder = ckb_vm::DefaultMachineBuilder::new(core_machine)
         .instruction_cycle_func(Box::new(instruction_cycles));
+    let machine_builder = machine_builder;
```
Comment (XuJiandong) [inline examples/ckb-vm-runner.rs]: not needed.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: tests/test_dy_memory.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        1024 * 1024 * 2,
+        format!("tests/programs/alloc_many"),
+        format!("alloc_many"),
+    );
+}
+
+#[test]
+fn test_memory_out_of_bounds() {
+    let memory_size = 1024 * 256;
+    let buffer = fs::read("tests/programs/alloc_many").unwrap().into();
```
Comment (XuJiandong) [inline tests/test_dy_memory.rs:50]: where is the source code of "alloc_many"?

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Review by mohanson — DISMISSED** [review]
LGTM

**Review by mohanson — APPROVED** [review]

**Pull request merged**